src/lib.rs

```diff
@@ -0,0 +1,230 @@
+extern crate ini;
+extern crate regex;
+extern crate semver;
+
+use ini::Ini;
+use regex::Regex;
+use std::default::Default;
+use std::fmt::{self, Display, Formatter};
+use std::path::Path;
+use std::str::FromStr;
+
+#[derive(Default)]
+pub struct AppVersion {
+    pub build_id: Option<String>,
+    pub code_name: Option<String>,
+    pub version_string: Option<String>,
+    pub source_repository: Option<String>,
+    pub source_stamp: Option<String>
+}
+
+impl AppVersion {
+    pub fn new() -> AppVersion {
+        Default::default()
+    }
+
+    fn update_from_application_ini(&mut self, ini_file: &Ini) {
+        if let Some(section) = ini_file.section(Some("App")) {
+            if let Some(build_id) = section.get("BuildID") {
+                self.build_id = Some(build_id.clone());
+            }
+            if let Some(code_name) = section.get("CodeName") {
+                self.code_name = Some(code_name.clone());
+            }
+            if let Some(version) = section.get("Version") {
+                self.version_string = Some(version.clone());
+            }
+            if let Some(source_repository) = section.get("SourceRepository") {
+                self.source_repository = Some(source_repository.clone());
+            }
+            if let Some(source_stamp) = section.get("SourceStamp") {
+                self.source_stamp = Some(source_stamp.clone());
+            }
+
+        }
+    }
+
+    fn update_from_platform_ini(&mut self, ini_file: &Ini) {
+        if let Some(section) = ini_file.section(Some("Build")) {
+            if let Some(build_id) = section.get("BuildID") {
+                self.build_id = Some(build_id.clone());
+            }
+            if let Some(version) = section.get("Milestone") {
+                self.version_string = Some(version.clone());
+            }
+            if let Some(source_repository) = section.get("SourceRepository") {
+                self.source_repository = Some(source_repository.clone());
+            }
+            if let Some(source_stamp) = section.get("SourceStamp") {
+                self.source_stamp = Some(source_stamp.clone());
+            }
+        }
+    }
+
+    pub fn version(&self) -> Option<Version> {
+        self.version_string
+            .as_ref()
+            .and_then(|x| Version::from_str(&*x).ok())
+    }
+}
+
+#[derive(Default)]
+pub struct Version {
+    major: u64,
+    minor: u64,
+    patch: u64,
+    pre: Option<(String, u64)>,
+}
+
+impl Version {
+    pub fn from_str(version_string: &str) -> Result<Version, &'static str> {
```

Comment:
Depending on how much work you want to do on this, I think it would make sense to define your own `Result` since this is a library.

Suggestion OTOH:

```rust
pub type Result<T> = result::Result<T, Error>;

#[derive(Debug)]
pub enum Error {
    MissingMajorVersion,
    MissingMinorVersion,

    Parse(ParseError),
    Io(io::Error),
}

#[derive(Clone, Debug)]
pub enum ParseError {
    PreleaseNumber,
    VersionString,
    SemVer(semver::ReqParseError),
}
```

Don’t let this be a blocker.